src/clojure/nrepl/socket.clj

```diff
@@ -0,0 +1,222 @@
+    :junixsocket
+    (Reflector/invokeConstructor junixsocket-address-class
+                                 (into-array File [(File. path)]))
+    (let [msg "Support for filesystem sockets requires JDK 16+ or a junixsocket dependency"]
+      (log msg)
+      (throw (ex-info msg {:nrepl/kind ::no-filesystem-sockets})))))
+
+(set! *warn-on-reflection* false)
+
+(defn- bind [sock addr] (.bind sock addr))
+(defn- get-path [addr] (.getPath addr))
+
+(set! *warn-on-reflection* orig-warn-on-reflection)
+
+(defn unix-server-socket
+  "Returns a filesystem socket bound to the path if the JDK is version
+  16 or newer or if com.kohlschutter.junixsocket/junixsocket-core can
+  be loaded dynamically.  Otherwise throws the ex-info map
+  {:nrepl/kind ::no-filesystem-sockets}."
+  [^String path]
+  (let [addr (unix-socket-address path)]
+    (case unix-domain-flavor
+      :jdk
+      (let [protocol (Reflector/getStaticField StandardProtocolFamily "UNIX")
+            sock (Reflector/invokeStaticMethod jdk-unix-server-class "open"
+                                               ^"[Ljava.net.StandardProtocolFamily;"
+                                               (into-array [protocol]))]
+        (bind sock addr)
+        (let [^Path path (get-path addr)]
+          (-> path .toFile .deleteOnExit))
+        sock)
+
+      :junixsocket
+      (let [new-instance (.getDeclaredMethod junixsocket-server-class
+                                             "newInstance" nil)
+            sock (.invoke new-instance nil nil)]
+        (bind sock addr)
+        (let [^String path (get-path addr)]
+          (-> path File. .deleteOnExit))
+        sock)
+
+      (let [msg "Support for filesystem sockets requires JDK 16+ or a junixsocket dependency"]
+        (log msg)
+        (throw (ex-info msg {:nrepl/kind ::no-filesystem-sockets}))))))
+
+(set! *warn-on-reflection* false)
+
+(defn as-nrepl-uri [sock transport-scheme]
+  (let [addr (and (some-> jdk-unix-server-class (instance? sock))
+                  (.getLocalAddress sock))]
+    (if (and addr (some-> jdk-unix-address-class (instance? addr)))
+      (URI. (str transport-scheme "+unix")
+            (-> sock .getLocalAddress .getPath .toAbsolutePath str)
+            nil)
+      (let [addr (and (some-> junixsocket-server-class (instance? sock))
+                      (.getLocalSocketAddress sock))]
+        (if (and addr (some-> junixsocket-address-class (instance? addr)))
+          (URI. (str transport-scheme "+unix")
+                (-> sock .getLocalSocketAddress .getPath)
+                nil)
+          ;; Assume it's an InetAddress socket
+          (URI. transport-scheme
+                nil  ;; userInfo
+                (-> sock .getInetAddress .getHostName)
+                (.getLocalPort sock)
+                nil  ;; path
+                nil  ;; query
+                nil))))))  ;; fragment
+
+(set! *warn-on-reflection* orig-warn-on-reflection)
+
+(defprotocol Acceptable
```

Comment:
Couldn't this very same technique be used to avoid reflection? defprotocol + extend-protocol for all relevant classes

Reply:
In some cases, but of course not where we don't know a suitable class will be available, either because earlier versions of the JVM didn't have it (e.g. (.getPath SocketAddress) before JVM 15), or because the relevant classes are optional and there's no useful parent (across all the supported JVM versions).

Prompted by your review, I was able to eliminate all of the `*warn-on-reflection*` manipulations except this one.  I'll push an update within the next few days.

Thanks much for the help.